packer/builder/azure/common/interruptible_task.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See the LICENSE file in the project root for license information.
+
+package common
+
+import (
+	"time"
+)
+
+type InterruptibleTaskResult struct {
+	Err         error
+	IsCancelled bool
+}
+
+type InterruptibleTask struct {
+	IsCancelled func() bool
+	Task        func() error
+}
+
+func NewInterruptibleTask(isCancelled func() bool, task func() error) *InterruptibleTask {
+	return &InterruptibleTask{
+		IsCancelled: isCancelled,
+		Task:        task,
+	}
+}
+
+func StartInterruptibleTask(isCancelled func() bool, task func() error) InterruptibleTaskResult {
+	t := NewInterruptibleTask(isCancelled, task)
+	return t.Run()
+}
+
+func (s *InterruptibleTask) Run() InterruptibleTaskResult {
+	completeCh := make(chan error)
+	defer close(completeCh)
+
+	go func() {
+		err := s.Task()
+		completeCh <- err
+	}()
+
+	cancelCh := make(chan InterruptibleTaskResult)
+	defer close(cancelCh)
+
+	go func() {
+		for {
+
+			if s.IsCancelled() {
+				cancelCh <- InterruptibleTaskResult{Err: nil, IsCancelled: true}
+				return
+			}
+
+			select {
+			case err := <-completeCh:
+				cancelCh <- InterruptibleTaskResult{Err: err, IsCancelled: false}
+				return
+			case <-time.After(100 * time.Millisecond):
+			}
+		}
```

Comment:
Might as well have this for loop running without the go-routine, since we'll block until we exit this loop anyway, right? Just return from the locations where cancelCh is being set now.